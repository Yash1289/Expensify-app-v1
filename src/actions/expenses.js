import { v4 as uuidv4 } from "uuid"
import database from "../firebase/firebase"

export const expenseAdder = (expense) => ({
    type: "ExpenseAdd",
    expense
});

export const startExpenseAdder = (expenseData = {}) => {
    return (dispatch) => {
        const { description = "", note = "", amount = 0, createdAt = 0 } = expenseData
        const expense = { description , note , amount , createdAt }
        return database.ref("expense").push(expense).then((snapshot) => {
                dispatch(expenseAdder({ id: snapshot.key, ...expense }))
        })
    }
}

 export const setExpense = (expenses) => ({
    type : "ExpenseSet",
    expenses
})

export const startSetExpense = () => {
    return(dispatch) => {
        return database.ref("expense").once("value").then((snapshot) => {
            const expenses = []
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id : childSnapshot.key ,
                    ...childSnapshot.val()
                })

            })
            dispatch(setExpense(expenses))
        })
    }
} 

export const expenseRemover = ({ id }) => ({
    type: "ExpenseRemove",
    id
})

 export const startExpenseRemover = (data) => {
    return(dispatch) => {
        const { id } = data
        return database.ref('expense').child(id).remove().then(() => {
            dispatch(expenseRemover(data))
        })
    }
} 

export const expenseEdit = (id, update) => ({
    type: "ExpenseEdit",
    id,
    update
})

 export const startExpenseEdit = (id , update) => {
    return (dispatch) => {
        return database.ref("expense").child(id).update({...update}).then(() => {
            dispatch(expenseEdit(id , update))
        })
    }
} 