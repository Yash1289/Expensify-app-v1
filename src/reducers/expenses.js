

const expenseReducerDefaultState = []

const expenseReducer = (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case "ExpenseAdd":
            return [
                ...state, action.expense
            ];
        case "ExpenseRemove":
            return state.filter((expense) => expense.id !== action.id)
        case "ExpenseEdit":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return { ...expense, ...action.update }
                }
                else {
                    return expense;
                }
            })
        default:
            return state;
    }
}

export default expenseReducer