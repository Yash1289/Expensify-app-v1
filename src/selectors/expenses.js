import moment from "moment"

const filterExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    const filteredExpenses = expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true ;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true ; 
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch;
    })

    return filteredExpenses.sort((a, b) => {
        if (sortBy === "date") {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        if (sortBy === "amount") {
            return a.amount < b.amount ? 1 : -1
        }
    })
}

export default filterExpenses