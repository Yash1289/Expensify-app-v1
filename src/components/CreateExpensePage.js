const React = require("react");
import ExpenseForm from "./ExpenseForm"
import { startExpenseAdder } from "../actions/expenses"
import { connect } from "react-redux"

export class CreateExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.CreateExpense(expense)
        this.props.history.push("/")
    }

    render(){
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => (
    {
        CreateExpense: (expense) => dispatch(startExpenseAdder(expense))
    }
)
export default connect(undefined , mapDispatchToProps)(CreateExpensePage)