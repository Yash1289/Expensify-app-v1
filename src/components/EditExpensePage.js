const React = require("react");
import { connect } from "react-redux"
import ExpenseForm from "./ExpenseForm"
import { startExpenseEdit , startExpenseRemover } from "../actions/expenses"

export class EditExpensePage extends React.Component { 
    onSubmit = (expense) => {
        this.props.ExpenseEdit(this.props.expense.id , expense)
        this.props.history.push("/")
    }
    onClick = () => {
        console.log(this.props.expense.id)
        this.props.ExpenseRemover({ id: this.props.expense.id })
        this.props.history.push("/")
    }
    render() {
        return (
            <div>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
                <button onClick={this.onClick} >Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state , props ) => {
    return {
        expense : state.expenses.find((expense) => expense.id === props.match.params.id )
    }
}

const mapDispatchToProps = ( dispatch , props ) => {
    return {
        ExpenseEdit: (id , expense) => dispatch(startExpenseEdit(id , expense)),
        ExpenseRemover : (data) => dispatch(startExpenseRemover(data))
    }
}

export default connect(mapStateToProps , mapDispatchToProps)(EditExpensePage) 