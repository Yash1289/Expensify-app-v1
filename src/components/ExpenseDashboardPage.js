const React = require("react");
import { connect } from "react-redux"
import { startSetExpense } from "../actions/expenses"
import ConnectedExpensesList from  "./ExpensesList"
import ExpensesListFilter from  "./ExpensesListFilter"
import ExpensesSummary from "./ExpenseSummary"


const ExpenseDashboardPage = () => (
     <div>
        <ExpensesSummary />
        <ExpensesListFilter />
        <ConnectedExpensesList />
    </div>
)



export default ExpenseDashboardPage
