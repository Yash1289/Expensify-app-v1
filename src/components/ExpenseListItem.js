import React from "react";
import {expenseRemover} from "../actions/expenses"
import { NavLink } from "react-router-dom"
import { connect } from "react-redux";
import moment from "moment"
const rupeeFormatter = require("rupee-formatter")

const ExpenseListItem = ({  id , description , amount, createdAt}) => (
    <div>
        <NavLink to={`/edit/${id}`} ><h3>{description}</h3></NavLink>
        <p>
        {rupeeFormatter(amount)} - {moment(createdAt).format("MMMM DD , YYYY")}
        </p>
    </div>
)



export default ExpenseListItem ;