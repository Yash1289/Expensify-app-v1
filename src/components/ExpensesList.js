import React from 'react'
import ExpenseListItem from "./ExpenseListItem"
import filterExpenses from "../selectors/expenses"
import { connect } from "react-redux"



export const ExpensesList = (props) => (
    <div>
        {props.expenses.length === 0 && <p>No Expenses</p>}
        {
            props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: filterExpenses(state.expenses, state.filter)
    }
}

const ConnectedExpensesList = connect(mapStateToProps)(ExpensesList)

export default ConnectedExpensesList