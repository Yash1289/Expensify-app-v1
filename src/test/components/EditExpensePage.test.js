import React from "react"
import { shallow } from "enzyme"
import { EditExpensePage } from "../../components/EditExpensePage"
import expenses from "../fixtures/expenses"

let ExpenseEdit , history , ExpenseRemover , wrapper ;
 beforeEach(() => {
      ExpenseEdit = jest.fn()
      history = {
         push: jest.fn()
     }
      ExpenseRemover = jest.fn()
      wrapper = shallow(<EditExpensePage
         expense={expenses[1]}
         ExpenseEdit={ExpenseEdit}
         history={history}
         ExpenseRemover={ExpenseRemover}
     />)
 }) 

test("Should render EditExpensePage correctly" ,() => {
    expect(wrapper).toMatchSnapshot()
})

test("Should call on submit with the necessary arguments", () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1])
    expect(ExpenseEdit).toHaveBeenLastCalledWith(expenses[1].id ,expenses[1])
    expect(history.push).toHaveBeenLastCalledWith("/")
})

test("Should call onClick with the necessary arguments ", () => {
    wrapper.find("button").simulate("click")
    expect(ExpenseRemover).toHaveBeenLastCalledWith({ id : expenses[1].id })
    expect(history.push).toHaveBeenLastCalledWith("/")
}) 