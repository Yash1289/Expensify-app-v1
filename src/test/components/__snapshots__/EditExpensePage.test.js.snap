// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 35,
        "createdAt": 345600000,
        "description": "Tea",
        "id": "2",
        "note": "Everyone drinks it so i drink it too what's the problem",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
